community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointerImpl.java

```diff
@@ -0,0 +1,143 @@
+{
+    private final LogFile logFile;
+    private final TransactionIdStore transactionIdStore;
+    private final CheckPointThreshold threshold;
+    private final StoreFlusher storeFlusher;
+    private final LogPruning logPruning;
+    private final KernelHealth kernelHealth;
+    private final Log msgLog;
+    private final CheckPointTracer tracer;
+
+    private WritableLogChannel channel;
+    private TransactionLogWriter transactionLogWriter;
+
+    public CheckPointerImpl( TransactionIdStore transactionIdStore,
+            CheckPointThreshold threshold, StoreFlusher storeFlusher, LogPruning logPruning, KernelHealth kernelHealth,
+            LogProvider logProvider, CheckPointTracer tracer, LogFile logFile )
+    {
+        this.logFile = logFile;
+        this.transactionIdStore = transactionIdStore;
+        this.threshold = threshold;
+        this.storeFlusher = storeFlusher;
+        this.logPruning = logPruning;
+        this.kernelHealth = kernelHealth;
+        this.msgLog = logProvider.getLog( CheckPointerImpl.class );
+        this.tracer = tracer;
+    }
+
+    @Override
+    public void start() throws Throwable
+    {
+        this.channel = logFile.getWriter();
+        this.transactionLogWriter =
+                new TransactionLogWriter( new LogEntryWriterV1( channel, NeoCommandHandler.EMPTY ) );
+    }
+
+    @Override
+    public void forceCheckPoint() throws IOException
+    {
+        doCheckPoint();
+    }
+
+    @Override
+    public void checkPointIfNeeded() throws IOException
+    {
+        if ( threshold.isCheckPointingNeeded() )
+        {
+            try ( LogCheckPointEvent event = tracer.beginCheckPoint() )
+            {
+                doCheckPoint();
+            }
+        }
+    }
+
+    private void doCheckPoint() throws IOException
+    {
+        long[] lastCommittedTransaction = transactionIdStore.getLastCommittedTransaction();
+        long lastCommittedTransactionId = lastCommittedTransaction[0];
+        LogPosition logPosition =
+                new LogPosition( lastCommittedTransaction[2], lastCommittedTransaction[3] );
+        msgLog.info( prefix( lastCommittedTransactionId ) + "Starting check pointing..." );
+
+        /*
+         * Check kernel health before going into waiting for transactions to be closed, to avoid
+         * getting into a scenario where we would await a condition that would potentially never
+         * happen.
+         */
+        kernelHealth.assertHealthy( IOException.class );
+
+        /*
+         * First we flush the store. If we fail now or during the flush, on recovery we'll find the
+         * earlier check point and replay from there all the log entries. Everything will be ok.
+         */
+        msgLog.info( prefix( lastCommittedTransactionId ) + " Starting store flush..." );
+        storeFlusher.forceEverything();
+
+        synchronized ( logFile )
+        {
+            synchronized ( channel ) // TODO: is this needed?
```

Comment:
This code is simply moved from somewhere else (I think I added it), and I _do_ think it's needed. Anyway that could be the topic of another, future, revisit.